### PR TITLE
Save special issues on special issue cancel and on associate EP

### PR DESCRIPTION
**Pull request by mdbenjam**
This change starts saving special issues when we force a user to cancel an EP because we can't handle the special issue. We also now save special issues when associating an EP. 

Connects: #811 

**Test Plan**
- [x] Manually check
  - [x] Select special issue that we cannot handle and ensure it's saved in the DB
  - [x] Select special issue and associate with an EP and ensure it's saved in the DB
- [x] Add new tests for the above to scenario
- [ ] Run test suite

### CODE REVIEW

File: app/controllers/establish_claims_controller.rb [hunk 1]
```diff
@@ -10,10 +10,21 @@ def perform
...
     task.assign_existing_end_product!(params[:end_product_id])
     render json: {}
   end
 
+  def cancel
+    task.appeal.update!(special_issues_params) if params[:special_issues]
+    task.cancel!(cancel_feedback)
+
+    respond_to do |format|
+      format.html { redirect_to establish_claims_path }
```
Comment (mdbenjam) [inline app/controllers/establish_claims_controller.rb]: Do we still need this?

File: client/app/containers/EstablishClaimPage/EstablishClaim.jsx [hunk 2]
```diff
@@ -204,13 +212,15 @@ export default class EstablishClaim extends BaseForm {
 
   handleCancelTask = () => {
     this.setState({
-      cancelModalDisplay: true
+      cancelModalDisplay: true,
+      handleFinishCancelTask: this.handleFinishCancelTask(false)
```
Comment (joofsh) [inline client/app/containers/EstablishClaimPage/EstablishClaim.jsx]: This feels awkward saving the whole cancel function in the state. What about just having a 
```
cancelModalDisplay: true,
submitSpecialIssuesOnCancel: true
```
And references that in the function?  Would that be cleaner than making a closure and saving it to the state?

File: client/app/containers/EstablishClaimPage/EstablishClaim.jsx [hunk 3]
```diff
@@ -347,15 +357,15 @@ export default class EstablishClaim extends BaseForm {
     let endProductInfo = this.getClaimTypeFromDecision();
 
 
-    return {
+    return ApiUtil.convertToSnakeCase({
```
Comment (joofsh) [inline client/app/containers/EstablishClaimPage/EstablishClaim.jsx]: nice! nested calls to this function lol

Comment (mdbenjam) [inline client/app/containers/EstablishClaimPage/EstablishClaim.jsx]: Maybe I should make this function recursive...

File: app/controllers/establish_claims_controller.rb [hunk 4]
```diff
@@ -4,13 +4,27 @@ class EstablishClaimsController < TasksController
...
-    task.assign_existing_end_product!(params[:end_product_id])
+    Task.transaction do
+      task.appeal.update!(special_issues_params)
+      task.assign_existing_end_product!(params[:end_product_id])
+    end
+    render json: {}
+  end
+
+  def cancel
+    Task.transaction do
```
Comment (mdbenjam) [inline app/controllers/establish_claims_controller.rb]: I added this as well as the previous methods to transaction blocks. I think it's safer that way.

File: client/app/util/ApiUtil.js [hunk 5]
```diff
@@ -30,7 +33,7 @@ const ApiUtil = {
         let snakeKey = key.replace(/([A-Z])/g, ($1) => `_${$1.toLowerCase()}`);
 
         // assign value to new object
-        result[snakeKey] = data[key];
+        result[snakeKey] = ApiUtil.convertToSnakeCase(data[key]);
```
Comment (joofsh) [inline client/app/util/ApiUtil.js]: minor nit, but better to call `this.convertToSnakeCase` (more flexible to name changes in the future)

File: client/app/util/ApiUtil.js [hunk 6]
```diff
@@ -22,6 +22,9 @@ const ApiUtil = {
 
   // Converts camelCase to snake_case
   convertToSnakeCase(data = {}) {
+    if (typeof data !== 'object') {
```
Comment (joofsh) [inline client/app/util/ApiUtil.js]: 😍

**Pull request merged**